capa/features/extractors/binexport2/__init__.py

```diff
@@ -362,16 +362,16 @@ def from_elf(cls, elf: ELFFile):
 
             regions.append(MemoryRegion(segment_rva, segment_data))
 
-        return cls(0, tuple(regions))
+        return cls(base_address, tuple(regions))
```

Comment:
@williballenthin let me know if there is a reason we were previously using `0` here for the base address - it was causing read errors that prevented `bytes` from being extracted properly. I've updated the code to use the base address retrieved from the BinExport file.

Reply:
In the existing design, an AddressSpace describes a module (PE/ELF) loaded into its preferred address. For PE, this is image base, and for ELF it's 0x0 because there's no preferred address. 

I think it's moderately important for the AddressSpace to depend only on the PE/ELF, to minimize the amount of context/concerns. It might be nice to be able to use the code in another project, for example. The changes you introduced seem to rely on the BinExport data, which I'm initially hesitant of. 

It might make more sense for the AddressSpace to use module RVAs everywhere, rather than preferred VAs. *shrug*

In any case, users of this class need to do a little address arithmetic before they attempt a read. This is expected. Maybe the documentation should be extended.  

I do recall I waffled a bit during the implementation of this class, and I'm open to further discussion. 

Thoughts?

Reply:
Perhaps we could introduce a method `rebase` that shifts an address space from one place to another. Then we can load from a PE/ELF to its preferred location and then rebase to what's found in the BinExport. Since it's only a matter of fixing up the range addresses, it's cheap.

Reply:
> Perhaps we could introduce a method `rebase` that shifts an address space from one place to another. Then we can load from a PE/ELF to its preferred location and then rebase to what's found in the BinExport. Since it's only a matter of fixing up the range addresses, it's cheap.

This sounds like a good option to me. My original change stemmed from capa not doing the arithmetic that you mentioned above before read resulting in read failures for ELF files (and inability to emit `bytes` features in most cases). Adding a `rebase` method would remove from `AddressSpace` users the burden of knowing what arithmetic and when to do it to help reduce bugs like this in the future.

Reply:
tracking in https://github.com/mandiant/capa/issues/2104